### DISCUSSION TITLE
Is it possible to scan only an area or a pixel ?

### DISCUSSION THREAD

**Discussion by sneakypandafr** [Spawnova/ShinsImageScanClass#24] [Ideas]
The scan of the whole window is making my computer lagging (because it's in a short loop).
Is it possible to change a part of the code to scan only an area or a pixel ?

Thanks for sharing this masterclass ! I found no equivalent

**Comment by Spawnova** [comment]
I think I added that already, but it's a bit wonky and only works with `AutoUpdate`  set to false

```ahk
f1::
scan.AutoUpdate := 0 ;have to manually update pixel data

scan.Update(50,50,100,100) ;instead of scanning the entire screen, look only in the region of 50,50 > 150,150

;because the pixel data is only reading the above region, technically all scans become region scans
;and some functions may not work as intended
if (scan.Image("image.png",0,x,y)) { 
    msgbox % "Found at " x "," y
}

scan.Update(90,200,20,20) ;scan for a pixel in the region of 90,200 > 110,220
if (scan.Pixel(0xFF0000,0,x,y)) {
    msgbox % "Pixel at " x "," y
}

scan.AutoUpdate := 1 ; reset if needed
return
```

It can be a bit confusing, I just haven't come up with a more elegant solution yet

**Reply by diogoml** [reply]
> I think I added that already, but it's a bit wonky and only works with `AutoUpdate` set to false
> 
> ```ahk
> f1::
> scan.AutoUpdate := 0 ;have to manually update pixel data
> 
> scan.Update(50,50,100,100) ;instead of scanning the entire screen, look only in the region of 50,50 > 150,150
> 
> ;because the pixel data is only reading the above region, technically all scans become region scans
> ;and some functions may not work as intended
> if (scan.Image("image.png",0,x,y)) { 
>     msgbox % "Found at " x "," y
> }
> 
> scan.Update(90,200,20,20) ;scan for a pixel in the region of 90,200 > 110,220
> if (scan.Pixel(0xFF0000,0,x,y)) {
>     msgbox % "Pixel at " x "," y
> }
> 
> scan.AutoUpdate := 1 ; reset if needed
> return
> ```
> 
> It can be a bit confusing, I just haven't come up with a more elegant solution yet

Hello.
Would it be possible to do this search starting from the center of the screen and searching in a circle until you find the image or restart the process?

**Reply by Spawnova** [reply]
You could try using ImageClosestToPoint(), it will return the image closest to the center of the region you are scanning
```ahk
f1::
scan.AutoUpdate := 0 ;have to manually update pixel data

scan.Update(50,50,500,500) ;instead of scanning the entire screen, look only in the region of 50,50 > 550,550

if (scan.ImageClosestToPoint("image.png",250,250,0,x,y)) { ;center would be 250,250 since the region is 500x500
    msgbox % "Found at " x "," y
}
scan.AutoUpdate := 1 ; reset if needed
return
```

**Reply by diogoml** [reply]
> You could try using ImageClosestToPoint(), it will return the image closest to the center of the region you are scanning
> 
> ```ahk
> f1::
> scan.AutoUpdate := 0 ;have to manually update pixel data
> 
> scan.Update(50,50,500,500) ;instead of scanning the entire screen, look only in the region of 50,50 > 550,550
> 
> if (scan.ImageClosestToPoint("image.png",250,250,0,x,y)) { ;center would be 250,250 since the region is 500x500
>     msgbox % "Found at " x "," y
> }
> scan.AutoUpdate := 1 ; reset if needed
> return
> ```

Hello,
I think I wasn't clear enough, sorry.
Anyway, for some reason, the code hasn't found the image.
I'll try to be clearer.
First point, is this library faster than the original AHK pixelsearch?
Within this library, which function would be more efficient, #ImageClosestToPoint or #PixelRegion?
I looked in the library instructions and found 8 different scanner direction modes, but none of them start from a central point, apparently.
My code would need the scan to start from a central point, like in the image below:
![SEARCH](https://github.com/user-attachments/assets/4eaf2154-0407-4b32-b00a-cb67e7daddbd)

In this mode, in theory, the scan would find the pixel or image in the order I put it.

Thanks for your help,

**Reply by Spawnova** [reply]
In desktop mode my library is slightly faster than ahk, in target window mode it is significantly faster and with specific use cases it's exponentially faster
There is no function for what you describe, instead you would probably want to use ImageArray, which will get all the found images then you can loop through them all to find their distances closest to a point, at which time you would know their order

**Reply by diogoml** [reply]
> In desktop mode my library is slightly faster than ahk, in target window mode it is significantly faster and with specific use cases it's exponentially faster There is no function for what you describe, instead you would probably want to use ImageArray, which will get all the found images then you can loop through them all to find their distances closest to a point, at which time you would know their order

Thanks for the reply.
I believe that in array mode the performance will drop significantly.
I really need to prioritize the speed of mouse clicks.
I'm having problems even using the basic script mode, apparently I'm not having success in finding the colors.
For example, I used the window mode to locate the colors inside "SEARCH.png", I used AHK's SPY to detect the color and wrote the script.
When I run it, however, it doesn't find the desired color.

```
`#noenv
setbatchlines,-1
settitlematchmode,1

;include the library assumed to be in the parent directory
;remove ../ if in same directory, or specify path
#include ShinsImageScanClass.ahk

scan := new ShinsImageScanClass("SEARCH.png") ;no title supplied so using desktop instead

F1::
;look for a pure red pixel anywhere on the desktop
if (scan.Pixel(0x3682FF,0,x,y)) {
	msgbox % "Found a red pixel at " x "," y
} else {
	msgbox % "Could not find a red pixel on the desktop!"
}


return

exitapp`

```
![print](https://github.com/user-attachments/assets/bd7b8857-6a68-44d5-ae8c-bb4bdaeb5522)

**Comment by sneakypandafr** [comment]
HUGE !!!!!
THANK YOU !
